### PR TITLE
Allow null-pointer-optimized enums in FFI if their underlying representation is FFI safe

### PR DESCRIPTION
**Pull request by mjbshaw**
I'm not sure if this requires an RFC. I attempted to start [a discussion on internals.rust-lang.org](https://internals.rust-lang.org/t/options-ffi-safety-and-guarantees-for-abi-compatibility-with-nonnull-optimizations/9784) and when no one really objected I figured I'd go ahead and try implementing this.

This allows types like `Option<NonZeroU8>` to be used in FFI without triggering the `improper_ctypes` lint. This works by changing the `is_repr_nullable_ptr` function to consider an enum `E` to be FFI-safe if:

- `E` has no explicit `#[repr(...)]`.
- It only has two variants.
- One of those variants is empty (meaning it has no fields).
- The other variant has only one field.
- That field is one of the following:
  - `&T`
  - `&mut T`
  - `extern "C" fn`
  - `core::num::NonZero*`
  - `core::ptr::NonNull<T>`
  - `#[repr(transparent)] struct` wrapper around one of the types in this list.
- The size of `E` and its field are both known and are both the same size (implying `E` is participating in the nonnull optimization).

This logic seems consistent with [the Rust nomicon](https://doc.rust-lang.org/nomicon/repr-rust.html).

### CODE REVIEW

File: src/librustc_lint/types.rs
Comment (Centril) [inline src/librustc_lint/types.rs]: Rewriting as `let data_idx = if ... { one } else if ... { zero } else { return false; }` would be more readable here.

File: src/librustc_lint/types.rs
Comment (Centril) [inline src/librustc_lint/types.rs]: You use `ty_def.variants[$idx].fields` a few times; refactor to lambda / function?

**Comment by Centril** [comment]
cc @rust-lang/wg-unsafe-code-guidelines

**Comment by rkruppe** [comment]
I do think we should change the lint to accepting types that are already effectively guaranteed to participate in Option layout optimizations (at minimum, the `NonZero*` types and `repr(transparent)` newtypes). However, there are also situations where rustc does apply layout optimizations but we don't obviously (i.e., without an RFC) want to commit to those optimizations.

Relatedly, there may be `repr(C)` library-defined types which happen to have a niche in a private field but whose author didn't intend to *guarantee* that `Option<Foo>` is FFI-safe. For example, a library author may at first use `bool` for a private field but later decide to use `uint8_t` to be defensive against the C side storing a byte other than 0 or 1, without expecting this to break clients.

So I think the check "`Option<T>` is the same size as `T`" that this PR currently implements is too broad. I would be more comfortable to have us start out with something more restrictive. I am not sure what that should be, though -- that uncertainty might be why this generalization hasn't happened yet.

**Comment by mjbshaw** [comment]
@Centril Thanks; I've made the changes you recommended. Let me know if there are others.

@rkruppe I agree. I've revised the code to now only allow:

- `&T`
- `extern "C" fn`
- `core::num::NonZero*`
- `core::ptr::NonNull<T>`
- `#[repr(transparent)] struct` wrappers around of of the types in this list.

I'll update the top-level comment to reflect this change.

File: src/librustc_lint/types.rs
Comment (rkruppe) [inline src/librustc_lint/types.rs]: I think now that the cases where non-null optimization is guaranteed are explicitly enumerated, this check should always pass if `ty_is_known_nonnull` returns true. After all, if the layout is unknowable (due to generics or whatever) then we can't very well claim that the non-null optimization will actually happen.

If that seems correct to you, then I would prefer this size comparison be converted into an assertion, so that we learn (from an ICE) if for some reason it turns out to be incorrect.

**Review by hanna-kruppe — CHANGES_REQUESTED** [review]

File: src/librustc_lint/types.rs
Comment (mjbshaw) [reply src/librustc_lint/types.rs]: Yes, this line is only needed because of RFC 2645, but my reasons for including it are a bit different.

RFC 2645 leaves it unspecified whether transparent `union`s participate in nonnull optimizations, and I don't intend to change that in this pull request. Continuing to emit the `improper_ctypes` warning in this situation makes it clear that the this is (still) unspecified.

Transparent `enum`s should be fine, though, since we can always safely optimize them just like a transparent `struct`. The feature will be gated in other parts of the code, and I don't think an additional gate check is necessary here for transparent `enum`s.

If people prefer, I can remove this `union` check from this pull request and move it into the implementation of RFC 2645.

File: src/librustc_lint/types.rs
Comment (mjbshaw) [reply src/librustc_lint/types.rs]: > Is there a specific reason why it's used here?

Because I don't know what I'm doing 😛. I used `reveal_all()` because 3 out of 4 places that involved getting the field type used it:

- Doesn't use `reveal_all`: [1](https://github.com/rust-lang/rust/blob/6cc24f26036b28fb3366de86efe3da6c4464057a/src/librustc_lint/types.rs#L454).
- Uses `reveal_all`: [1](https://github.com/rust-lang/rust/blob/6cc24f26036b28fb3366de86efe3da6c4464057a/src/librustc_lint/types.rs#L514-L517), [2](https://github.com/rust-lang/rust/blob/6cc24f26036b28fb3366de86efe3da6c4464057a/src/librustc_lint/types.rs#L562-L565), [3](https://github.com/rust-lang/rust/blob/6cc24f26036b28fb3366de86efe3da6c4464057a/src/librustc_lint/types.rs#L514-L517).

If I shouldn't be doing this (or if I should be doing it only once but somewhere else, like in `is_repr_nullable_ptr`), please let me know!

File: src/librustc_lint/types.rs
Comment (mjbshaw) [reply src/librustc_lint/types.rs]: Done.

File: src/librustc_lint/types.rs
Comment (mjbshaw) [reply src/librustc_lint/types.rs]: > Can you please explain why it's needed here and not there?

I wish I could, but I must admit this was a cargo cult copy 'n' paste of [items mentioned in my other comment](https://github.com/rust-lang/rust/pull/60300#discussion_r280157289). I don't understand when `field.ty(cx, substs)` is sufficient or when it needs to be wrapped in `tcx.normalize_erasing_regions(..)`. I would really appreciate it if someone could educate me on this part. I'm happy to change the code to just be `field.ty(cx, substs)` if that's what should be done.

**Comment by joshtriplett** [comment]
The idea of this change looks *great* to me. We want to define this behavior, carefully, and not leave it undefined. And this allows people to define *safer* types for FFI functions.

We do need further review on the code itself, but let's get consensus on the *concept*:

@rfcbot merge

**Comment by scottmcm** [comment]
:+1: now that this is the "careful" version (following https://github.com/rust-lang/rust/pull/60300#issuecomment-487000474)

File: src/librustc_lint/types.rs
Comment (rkruppe) [reply src/librustc_lint/types.rs]: Ok, this makes sense. I think I'd prefer to only accept structs here for now and cover other transparent ADTs once they actually exist. You're right that it should probably Just Work for enums, but I'd like to have tests for that and prefer to not have dead, untestable code, even temporarily.

File: src/librustc_lint/types.rs
Comment (rkruppe) [reply src/librustc_lint/types.rs]: I am not quite sure either. It makes intuitive sense to me that *normalization* is needed in all of those places (including the one which currently isn't normalizing). But after some digging I am pretty sure that the use of reveal_all specifically in the normalization leads to undesirable behavior. 

For example, in [this program](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=76fdb36dcf516354a400647c26b64e87) using existential types, it makes the lint pierce through otherwise-opaque existentials and both
- accept code that is only FFI-safe because of the particular type underlying the existential, which users of that type shouldn't know or rely on
- leak the underlying type to the user in error messages when (correctly) rejecting uses of an existential type as FFI-unsafe

I believe if we didn't use reveal_all in those places (but just normalized with a `UserFacing` reveal mode) the lint would complain about both of those functions in terms of the existential types. IMO that is clearly better behavior, but this is also a pre-existing issue, so I'm not sure how to best proceed.

**Comment by pnkfelix** [comment]
... why isn't there the standard list of checkboxes for all members of labelled teams (T-lang in this case) on the [rfcbot comment](https://github.com/rust-lang/rust/pull/60300#issuecomment-488798080)

File: src/librustc_lint/types.rs
Comment (mjbshaw) [reply src/librustc_lint/types.rs]: Okay, I've modified this to only accept `struct`s for now. I'll revise this later to accept transparent `enum`s (either as a part of pull request #60463 (if this pull request is merged first), or as a later update to this pull request (if #60463 is merged first)).

**Comment by nikomatsakis** [comment]
@pnkfelix perhaps it was slow -- the list exists now

**Comment by gnzlbg** [comment]
> One of those variants is empty.

Here empty means that the variant has no fields, right? That is, variants with zero-sized fields are not empty, right? That is, `Result<(), NonZeroU8>` is an improper ctype. If so, is there a test covering this?

**Comment by mjbshaw** [comment]
@gnzlbg Good idea. I added one. And yes, "empty" means "no fields". I'll clarify the top post to mention that.

@rkruppe I appreciate your review from earlier, but there are a couple items left I'm not sure how to address. Should I just try changing `tcx.normalize_erasing_regions(...)` to `field.ty(cx, substs)` and test if it works, and if so, assume it's enough?

File: src/librustc_lint/types.rs
Comment (rkruppe) [inline src/librustc_lint/types.rs]: Please add a message so that if someone trips this ICE and goes to file a bug, there's a recognizable panic message to search for. Maybe something like

> improper_ctypes: Option nonnull optimization not applied?

**Review by hanna-kruppe — CHANGES_REQUESTED** [review]

**Comment by rkruppe** [comment]
I think such a change would be good, but it would deserve more thorough vetting (e.g., should have a reviewer who actually understands normalization and reveal modes) and I don't want to hold up this PR that long. I'll file an issue for it, and if you fix my last nit then r=me when FCP ends.

**Comment by rkruppe** [comment]
> if you fix my last nit then r=me when FCP ends.

Sorry to only notice now, but before merging there are some procedural issues to resolve:

- this branch has merge commits, that's generally discouraged, please rebase (and while at it, you could squash some back-and-forth in the history, but no big deal) 
- the new rustc_* attribute doesn't have a feature gate test, please add one

**Comment by pnkfelix** [comment]
switching status to S-waiting-on-author to reflect the need to address the changes noted by @rkruppe above.

File: src/librustc_lint/types.rs
Comment (mjbshaw) [reply src/librustc_lint/types.rs]: Done.

Comment (rkruppe) [reply src/librustc_lint/types.rs]: Are you sure? Is Github's UI playing tricks on me?

**Comment by mjbshaw** [comment]
> this branch has merge commits, that's generally discouraged, please rebase (and while at it, you could squash some back-and-forth in the history, but no big deal)

Thanks, I wasn't sure what Rust's policy was on this. I'll keep that in mind going forward (also: done).

> the new rustc_* attribute doesn't have a feature gate test, please add one

Done.

File: src/librustc_lint/types.rs
Comment (mjbshaw) [reply src/librustc_lint/types.rs]: Sorry, I might not have pushed by the time I pressed the "comment" button. It's done for realsies this time.

**Comment by rkruppe** [comment]
Thank you for your work and patience!

@bors r+

File: src/libcore/num/mod.rs
Comment (oli-obk) [inline src/libcore/num/mod.rs]: Could this just reuse the attribute in the line above?

Comment (RalfJung) [reply src/libcore/num/mod.rs]: I think the goal was to separate "performing the optimization" from "guaranteeing the optimization".

**Pull request merged**
